File: packages/guards/src/geographic-bias.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface GeographicBiasOptions { action: 'block' | 'warn'; }

const BIAS_PATTERNS: RegExp[] = [
  /\b(?:third[- ]world|backward|uncivilized|primitive)\s+(?:country|countries|nation|people)\b/gi,
  /\b(?:developed|civilized|advanced)\s+(?:world|countries|nations?)\s+(?:vs|versus|compared to)\b/gi,
  /\b(?:those|these)\s+(?:people|countries)\s+(?:in|from)\s+(?:Africa|Asia|Latin America|Middle East)\b/gi,
  /\b(?:typical|always|never)\s+(?:in|for)\s+(?:Africa|Asia|Eastern Europe|Latin America)\b/gi,
  /\b(?:western|American|European)\s+(?:values?|standards?|way)\s+(?:is|are)\s+(?:better|superior|correct)\b/gi,
  /\b(?:shithole|s-hole)\s+(?:country|countries)\b/gi,
];

export function geographicBias(options: GeographicBiasOptions): Guard {
  return { name: 'geographic-bias', version: '0.1.0', description: 'Detect geographic and cultural bias', category: 'content', supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now(); const matched: string[] = [];
      for (const p of BIAS_PATTERNS) { const re = new RegExp(p.source, p.flags); const m = re.exec(text); if (m) matched.push(m[0]); }
      const triggered = matched.length > 0;
      return { guardName: 'geographic-bias', passed: !triggered, action: triggered ? options.action : 'allow',
        message: triggered ? `Geographic bias: "${matched[0]}"` : undefined,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { matched, reason: 'Text contains geographic or cultural bias' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/geographic_bias.py
"""Detect geographic stereotypes."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_GEO_BIAS_PATTERNS = [
    re.compile(
        r"people\s+from\s+\w+\s+are\s+always",
        re.IGNORECASE,
    ),
    re.compile(
        r"all\s+\w+(?:\s+people)?\s+are\s+",
        re.IGNORECASE,
    ),
    re.compile(
        r"typical\s+\w+\s+behavior",
        re.IGNORECASE,
    ),
    re.compile(
        r"(?:every|all)\s+(?:american|chinese|"
        r"indian|japanese|korean|mexican|"
        r"african|european|russian|brazilian|"
        r"british|french|german|italian|"
        r"arab|australian)s?\s+(?:are|always|"
        r"never|can't|cannot)",
        re.IGNORECASE,
    ),
    re.compile(
        r"those\s+\w+\s+(?:people|folks)\s+"
        r"(?:always|never|can't)",
        re.IGNORECASE,
    ),
]


class _GeographicBias:
    def __init__(
        self, *, action: str = "warn"
    ) -> None:
        self.name = "geographic-bias"
        self.action = action

    def check(
        self, text: str, stage: str = "input"
    ) -> GuardResult:
        start = time.perf_counter()
        found: list[str] = []

        for pat in _GEO_BIAS_PATTERNS:
            match = pat.search(text)
            if match:
                found.append(match.group())

        elapsed = (
            time.perf_counter() - start
        ) * 1000

        if not found:
            return GuardResult(
                guard_name="geographic-bias",
                passed=True,
                action="allow",
                latency_ms=round(elapsed, 2),
            )

        return GuardResult(
            guard_name="geographic-bias",
            passed=False,
            action=self.action,
            message=(
                "Geographic stereotype detected"
            ),
            latency_ms=round(elapsed, 2),
            details={
                "matched": found,
                "reason": (
                    "Text contains geographic"
                    " stereotyping language"
                ),
            },
        )


def geographic_bias(
    *, action: str = "warn"
) -> _GeographicBias:
    return _GeographicBias(action=action)
